File: 0152-maximum-product-subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        self.global_max = float('-inf')
        @cache
        def findMinMax(index):
            if index == len(nums) - 1:
                self.global_max = max(self.global_max, nums[index])
                return nums[index], nums[index]
            nextMin, nextMax = findMinMax(index + 1)
            cur = nums[index]
            
            curMin = min(cur, cur*nextMin, cur*nextMax)
            curMax = max(cur, cur*nextMin, cur*nextMax)
            self.global_max = max(self.global_max, curMax)
            return curMin, curMax
        findMinMax(0)
        return self.global_max

File: 0152-maximum-product-subarray/solution.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int minval = nums[0];
        int maxval = nums[0];
        int res = nums[0];
        for(int i = 1; i < nums.size(); ++i){
            if(nums[i] == 0){
                res = max(res, nums[i]);
                minval = 1;
                maxval = 1;
                continue;
            }
            int cur = nums[i];
            int temp = maxval;
            maxval = max(maxval * cur, minval * cur);
            maxval = max(maxval, cur);

            minval = min(temp * cur, minval * cur);
            minval = min(minval, cur);
            res = max(res, maxval);
        }
        return res;
    }
};
